app/services/hashed_password.rb

```diff
@@ -0,0 +1,75 @@
+# HashedPassword is returned from PasswordHasher.hash.
+# HashedPassword outputs various key hashing related to NIST encryption.
+#
+# Generate and store a 128-bit salt S.
+# Z1, Z2 = scrypt(S, password)   # split 256-bit output into two halves
+# Generate random R.
+# D = KMS_GCM_Encrypt(key=server_secret, plaintext=R) xor Z1
+# E = hash( Z2 + R )
+# F = hash(E)
+# Store F (User.encrypted_password) and D (User.encryption_key) in db
+
+class HashedPassword
+  attr_accessor :encrypted_d, :salt, :z1, :z2, :random_r
+
+  def initialize(salt:, z_one:, z_two:, random_r:)
+    self.salt = salt
+    self.z1 = z_one
+    self.z2 = z_two
+    self.random_r = random_r
+    @unlocked = false
+  end
+
+  def as_scrypt_hash
+    "#{salt}$#{z1}#{z2}"
+  end
+
+  def hash_e
+    OpenSSL::Digest::SHA256.hexdigest(z2 + random_r)
+  end
+
+  def hash_f
```

Comment:
I just kept reading and realized other files reference `hash_e` and stuff as well, so the comment doesn't really stand, we want multiple intermediate values throughout the app :/